bundles/org.openhab.binding.argoclima/src/main/java/org/openhab/binding/argoclima/internal/configuration/ArgoClimaConfigurationBase.java

```diff
@@ -0,0 +1,374 @@
+    /////////////////////
+
+    /**
+     * Utility function for logging only. Gets a parsed value from the supplier function or, exceptionally the raw
+     * value. Swallows exceptions.
+     *
+     * @param <T> Actual type of variable returned by the supplier (parsed)
+     * @param fn Parser function
+     * @return String param value (if parsed correctly), or the default value post-fixed with {@code [raw]} - on parse
+     *         failure.
+     */
+    protected static <@NonNull T> String getOrDefault(ConfigValueSupplier<T> fn) {
+        try {
+            return fn.get().toString();
+        } catch (ArgoConfigurationException e) {
+            return e.rawValue + "[raw]";
+        }
+    }
+
+    @Override
+    public final String toString() {
+        return String.format("Config: { %s, deviceCpuId=%s, refreshInterval=%d, oemServerPort=%d, oemServerAddress=%s,"
+                + "schedule1DayOfWeek=%s, schedule1OnTime=%s, schedule1OffTime=%s, schedule2DayOfWeek=%s, schedule2OnTime=%s, schedule2OffTime=%s, schedule3DayOfWeek=%s, schedule3OnTime=%s, schedule3OffTime=%s, resetToFactoryDefaults=%s}",
+                getExtraFieldDescription(), deviceCpuId, refreshInterval, oemServerPort,
+                getOrDefault(this::getOemServerAddress),
+                getOrDefault(() -> getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_1)),
+                getOrDefault(() -> getScheduleOnTime(ScheduleTimerType.SCHEDULE_1)),
+                getOrDefault(() -> getScheduleOffTime(ScheduleTimerType.SCHEDULE_1)),
+                getOrDefault(() -> getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_2)),
+                getOrDefault(() -> getScheduleOnTime(ScheduleTimerType.SCHEDULE_2)),
+                getOrDefault(() -> getScheduleOffTime(ScheduleTimerType.SCHEDULE_2)),
+                getOrDefault(() -> getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_3)),
+                getOrDefault(() -> getScheduleOnTime(ScheduleTimerType.SCHEDULE_3)),
+                getOrDefault(() -> getScheduleOffTime(ScheduleTimerType.SCHEDULE_3)), resetToFactoryDefaults);
+    }
+
+    /**
+     * Return derived class'es extra configuration parameters (for a common {@link toString} implementation)
+     *
+     * @return Comma-separated list of configuration parameter=value pairs or empty String if derived class does not
+     *         introduce any.
+     */
+    protected abstract String getExtraFieldDescription();
+
+    /**
+     * Validate derived configuration
+     *
+     * @throws ArgoConfigurationException - on validation failure
+     */
+    protected abstract void validateInternal() throws ArgoConfigurationException;
+
+    /**
+     * Validate current config
+     *
+     * @return Error message if config is invalid. Empty string - otherwise
+     */
+    public final String validate() {
+        if (refreshInterval < 0) {
+            return "Refresh interval must be >= 0";
+        }
+
+        if (oemServerPort < 0 || oemServerPort >= 65536) {
```

Comment:
See other comments, please check if it should be equal or greather then. Atleast the returned string does not match this condition

Reply:
FYI - XML was wrong. The range is `0..65535` (inclusive)

Reply:
I think this has been changed in prev. commit already. Let me know if I missed anything